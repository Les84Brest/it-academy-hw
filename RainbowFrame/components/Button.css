.btn{
  font: 17px/27px Arial, sans-serif;
  background-color: #f57413;
  cursor: pointer;
  background: -webkit-linear-gradient(top,#ed8414, #f57413);
  background: -moz-linear-gradient(top,#ed8414, #f57413);
  background: -ms-linear-gradient(top,#ed8414, #f57413);
  background: -o-linear-gradient(top,#ed8414, #f57413);
  padding: 2px 7px;
  margin: 10px 0 5px;
  position: relative;
  box-shadow: -webkit-box-shadow -1px 1px 2px rgba(0,0,0,0.28), inset 0 1px 0 0 #ffa854;
  box-shadow: -moz-box-shadow -1px 1px 2px rgba(0,0,0,0.28),inset 0 1px 0 0 #ffa854;
  box-shadow: -1px 1px 1px rgba(0,0,0,0.28), 	inset 0 1px 0 0 #ffa854;
  border: 1px solid #d86c1b;
  display: block;
  text-transform: uppercase;
  color: #fff;
  min-width: 107px;
  text-shadow: 1px 1px 1px rgba(0,0,0, 0.3);
  line-height: 40px;
  text-decoration: none;
  border-radius: 5px;
  vertical-align: middle;
  text-align: center;
  }

.btn:hover{
  background: #ff911b;
}
.btn:active{
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.06);
  transform: scale(.9,.9);
  -moz-transform: scale(.9,.9);
  -webkit-transform: scale(.9,.9);
  -o-transform: scale(.9,.9);
  -ms-transform: scale(.9,.9);
  background: #ff911b;
}
.btn:disabled{
  opacity: 0.3;
}