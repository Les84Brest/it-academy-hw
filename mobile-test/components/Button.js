
import React, {Children, Component} from "react";
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

// import './Button.css';



const Button = ({children, onClick, className, disabled, active, ...attrs}) => {
  
  const classes = ClassNames(
    'btn',
    className,
    {active},
  );

  const onClickAction = e => {
    if(disabled){
      e.preventDefault();
    }else{
      return onClick(e);
    }
  }
    const Tag = attrs.href ? 'a' : 'button';
  return(
    <Tag
    {...attrs}
    className={classes}
    onClick={onClickAction}
    disabled={disabled}
    >{children}</Tag>
  );
}

Button.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  active: PropTypes.bool,
};

Button.defaultProps = {
  children: 'DefaultButton',
  onClick : () => {},
  className : '',
  disabled: false,
  active: false,
}
export default Button;
