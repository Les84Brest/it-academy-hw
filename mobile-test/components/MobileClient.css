 tr.MobileClient td {
   border-right: 1px solid #dfdfdf;
 }

 tr.MobileClient td:last-child {
   border-right: 0px;
 }

 tr.MobileClient td.active{
   background-color: darkgreen;
 }
 tr.MobileClient td.blocked{
   background-color: tomato;
 }
/* добавление нового клиента */
 .MobileClient-new{
   color: tomato;
 }