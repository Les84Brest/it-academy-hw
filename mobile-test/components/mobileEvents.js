import EventEmmiter from 'events';

var mobileEvents = new EventEmmiter();

export default mobileEvents;