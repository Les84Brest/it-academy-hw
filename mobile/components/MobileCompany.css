.MobileCompany table {
  width: 100%;
  text-align: center;
  border-bottom: 2px solid #dfdfdf;
  border-radius: 6px;
  border-collapse: separate;
  border-spacing: 0px;
}

.MobileCompany {
  padding: 15px 5px;
}
.MobileCompany table thead tr {
  color: #ffffff;
  font-weight: bold;
  background: #c83240;
}
.MobileCompany table thead tr td {
  padding: 10px 5px;
}


.MobileCompany table tbody tr:nth-child(1n) {
  background: #f6f6f6;
}

.MobileCompany table tbody tr:nth-child(2n) {
  background: #e6e6e6;
}
.section{
  padding-top: 10px;
  padding-bottom: 10px;
  border-bottom: 1px solid darkcyan;
}
