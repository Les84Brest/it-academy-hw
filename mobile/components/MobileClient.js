import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Button from './Button';

import mobileEvents from './mobileEvents';

import './MobileClient.css';



class MobileClient extends React.PureComponent {

  static WORKMODE_SHOW = 1;
  static WORKMODE_EDIT = 2;
  static WORKMODE_NEW = 3;

  constructor(props) {
    super(props);
    // управляем режимом компонента


    if (props.clientInfo == null) {
      this.state.workMode = MobileClient.WORKMODE_NEW;
      // build new client hash for state

      this.state.clientInfo = {
        lastName: '',
        firstName: '',
        secondName: '',
        balanse: '',
        id: Math.floor(Math.random() * 100000),
        status: '',
      };
    }

    //refs

    this.firstNameRef = null;
    this.lastNameRef = null;
    this.secondNameRef = null;
    this.balanseRef = null;
    
    //new client
    this.firstNameNewRef = null;
    this.lastNameNewRef = null;
    this.secondNameNewRef = null;
    this.balanseNewRef = null;
    

  }

  // ref callbacks


  setLastNameRef = (ref) => {
    this.lastNameRef = ref;
  }
  setSecondNameRef = (ref) => {
    this.secondNameRef = ref;
  }
  setFirstNameRef = (ref) => {
    this.firstNameRef = ref;
  }
  setBalanseRef = (ref) => {
    this.balanseRef = ref;
  }
// new client ref
  setLastNameNewRef = (ref) => {
    this.lastNameNewRef = ref;
  }
  setSecondNameNewRef = (ref) => {
    this.secondNameNewRef = ref;
  }
  setFirstNameNewRef = (ref) => {
    this.firstNameNewRef = ref;
    console.log(this.firstNameNewRef); 
  }
  setBalanseNewRef = (ref) => {
    console.log("ref balanse new");
    this.balanseNewRef = ref;
    console.log(this.balanseNewRef);
  }


  static propTypes = {

    clientInfo: PropTypes.shape({
      lastName: PropTypes.string,
      firstName: PropTypes.string,
      secondName: PropTypes.string,
      balanse: PropTypes.number,
      id: PropTypes.number,
      status: PropTypes.string,
    }
    )

  }

  state = {
    clientInfo: this.props.clientInfo,
    workMode: MobileClient.WORKMODE_SHOW,
  }


  componentDidUpdate = () =>{
    
    this.setState({clientInfo: this.props.clientInfo});
  }

  handlerSaveNewClient = () => {

    let client = { ...this.state.clientInfo };
   console.log(this.balanseNewRef.value); 

    client.lastName = this.lastNameNewRef.value;
    client.firstName = this.firstNameNewRef.value;
    client.secondName = this.secondNameNewRef.value;
    client.balanse = parseFloat(this.balanseNewRef.value);

    if (client.balanse <= 0) {
      client.status = 'blocked';
    } else {
      client.status = 'active';
    }

    mobileEvents.emit('savingClient', client);
    //переключаем отображение на показ 
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });
  }

  handlerSaveClient = () => {
    //копия клиента из state
    let client = { ...this.state.clientInfo };


    client.lastName = this.lastNameRef.value;
    client.firstName = this.firstNameRef.value;
    client.secondName = this.secondNameRef.value;
    client.balanse = parseFloat(this.balanseRef.value);

    if (client.balanse <= 0) {
      client.status = 'blocked';
    } else {
      client.status = 'active';
    }

    mobileEvents.emit('savingClient', client);
    //переключаем отображение на показ 
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });

  }

  handlerEditClient = () => {
    this.setState({ workMode: MobileClient.WORKMODE_EDIT });
  }
  handlerEditCancel = () => {
    this.setState({ workMode: MobileClient.WORKMODE_SHOW });
  }

  setTextInput = (element) => {
    this.newTextRef = ref;
  }


  render() {
    console.log('MobileClient render. ID - ' + this.state.clientInfo.id);


    return (
      <Fragment>
        {

          (this.state.workMode == MobileClient.WORKMODE_SHOW) &&
          <tr className='MobileClient'>
            <td >{this.state.clientInfo.lastName}</td>
            <td >{this.state.clientInfo.firstName}</td>
            <td >{this.state.clientInfo.secondName}</td>
            <td >{this.state.clientInfo.balanse}</td>
            <td className={classNames({ active: this.state.clientInfo.balanse > 0, blocked: this.state.clientInfo.balanse <= 0 })} >{this.state.clientInfo.status}</td>
            <td><Button onClick={this.handlerEditClient}>Редактировать</Button></td>
            <td><Button onClick={()=>{mobileEvents.emit('deletingClient', this.state.clientInfo.id)}}>Удалить</Button></td>
          </tr>

        }
        {
          (this.state.workMode == MobileClient.WORKMODE_EDIT) &&
          <tr className='MobileClient'>
            <td><input type="text" ref={this.setLastNameRef} defaultValue={this.state.clientInfo.lastName} /></td>
            <td><input type="text" ref={this.setFirstNameRef} defaultValue={this.state.clientInfo.firstName} /></td>

            <td><input type="text" ref={this.setSecondNameRef} defaultValue={this.state.clientInfo.secondName} /></td>
            <td><input type="number" ref={this.setBalanseRef} defaultValue={parseFloat(this.state.clientInfo.balanse)} /></td>
            <td> </td>
            <td><Button onClick={this.handlerEditCancel}>Отмена</Button></td>
            <td><Button onClick={this.handlerSaveClient}>Сохранить</Button></td>
          </tr>
        }
        {

          (this.state.workMode == MobileClient.WORKMODE_NEW) &&
          <div className="MobileClient-new">
            <label className="MobileClient-new__input">Фамилия
              <input type="text" ref={this.setLastNameNewRef} defaultValue="" />
            </label>
            <label className="MobileClient-new__input">Имя
              <input type="text" ref={this.setFirstNameNewRef}  defaultValue="" />
            </label>
            <label className="MobileClient-new__input">Отчество
              <input type="text" ref={this.setSecondNameNewRef}  defaultValue="" />
            </label>
            <label  className="MobileClient-new__input">Баланс
              <input type="number" ref={this.setBalanseNewRef}  defaultValue="" />
            </label>
            <br />
            <Button onClick={this.handlerSaveNewClient}>Сохранить</Button>
            <Button onClick={()=>{mobileEvents.emit('addNewClientCancel')}}>Отмена</Button>
          </div>
        }
      </Fragment>

    )
  }
}

export default MobileClient;