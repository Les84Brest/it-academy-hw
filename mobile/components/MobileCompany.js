import React from 'react';
import PropTypes from 'prop-types';
import MobileClient from './MobileClient';
import Button from './Button';
import EventEmmiter from 'events';

import './MobileCompany.css';




class MobileCompany extends React.PureComponent {

  constructor(props) {
    super(props);
    

    //события
    this.mobileEvents = new EventEmmiter();
    this.mobileEvents.on('filterShowAll', this.handlerFilterDisplayAll);
    this.mobileEvents.on('filterShowActive', this.handlerFilterDisplayActive);

  }

  

  static propTypes = {
    companyData: PropTypes.objectOf(
      PropTypes.arrayOf(
        PropTypes.shape({
          lastName: PropTypes.string,
          firstName: PropTypes.string,
          secondName: PropTypes.string,
          balanse: PropTypes.number,
          id: PropTypes.number,
          status: PropTypes.string,
        }
        )
      )
    )
  }

  state = {
    currentCompany: "MTS",
    clients: this.props.companyData.MTS,

  }

  handlerChangeCompany = (e) => {
    e.preventDefault();
    
    let newProps = {
      currentCompany: e.target.dataset.company,
      clients: this.props.companyData[e.target.dataset.company],
    }
    this.setState(newProps);
  }
  handlerFilterDisplayActive = () => {
    let activeClients=[...this.props.companyData[this.state.currentCompany]].filter(item => {
      if (item.status == 'active') {
        return item;
      }
      this.setState({clients: activeClients});
    });

  }
  handlerFilterDisplayAll = () => {
    let allClients = [...this.props.companyData[this.state.currentCompany]];
    this.setState({clients: allClients});
    console.log(allClients);
  }
 

  render() {
    console.log('MobileCompany render');

    let clientsList = this.state.clients.map(item => <MobileClient key={item.id} clientInfo={item} />);
    return (
      <div className="MobileCompany">
        <div className="MobileCompany__header section">
          <div className="MobileCompany__company-change section">
            <Button onClick={this.handlerChangeCompany} data-company="MTS">MTS</Button>
            <Button onClick={this.handlerChangeCompany} data-company="velcom">Velcom</Button>
            <h2>{`Название компании - ${this.state.currentCompany}`}</h2>
          </div>
          <div className="MobileCompany__filter section">
            <Button onClick={() => {this.mobileEvents.emit('filterShowAll')}}>Все</Button>
            <Button>Активные</Button>
            <Button>Заблокированные</Button>

          </div>

        </div>
        <table>
          <thead>
            <tr>
              <td>Фамилия</td>
              <td>Имя</td>
              <td>Отчество</td>
              <td>Баланс</td>
              <td>Статус</td>
              <td>Редактировать</td>
              <td>Удалить</td>

            </tr>
          </thead>
          <tbody>

            {clientsList}

          </tbody>
        </table>
        <Button>Добавить нового</Button>
      </div>
    )
  }
}

export default MobileCompany;