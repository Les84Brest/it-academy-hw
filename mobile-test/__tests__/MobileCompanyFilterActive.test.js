"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';
import Button from "../components/Button";
import MobileClient from '../components/MobileClient';
import mobileEvents from '../components/mobileEvents';

test('MobileCompany фильтр: показать активных', () => {
  // исходные данные для рендера компонента
  const data = [{
    "company": "MTS",
    "lastName": "Маховицкий",
    "firstName": "Эрнест",
    "secondName": "Афанасиевич",
    "balanse": 10.55,
    "id": 105,
    "status": "active"
  },
  {
    "company": "MTS",
    "lastName": "Набойщикова",
    "firstName": "Лиана",
    "secondName": "Агафоновна",
    "balanse": -3.5,
    "id": 125,
    "status": "blocked"
  },]
  // создаем экземпляр компонента для теста
  const component = renderer.create(
    <MobileClient clients={data} />
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
  let btn = component.root.find(el => {
    return el.props.children == "Активные";
  });
  
  btn.props.onClick();

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();



});