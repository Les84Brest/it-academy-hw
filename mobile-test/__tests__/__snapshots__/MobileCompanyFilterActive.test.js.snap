// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileCompany фильтр: показать активных 1`] = `
<div
  className="MobileClient-new"
>
  <label
    className="MobileClient-new__input"
  >
    Фамилия
    <input
      defaultValue=""
      type="text"
    />
  </label>
  <label
    className="MobileClient-new__input"
  >
    Имя
    <input
      defaultValue=""
      type="text"
    />
  </label>
  <label
    className="MobileClient-new__input"
  >
    Отчество
    <input
      defaultValue=""
      type="text"
    />
  </label>
  <label
    className="MobileClient-new__input"
  >
    Баланс
    <input
      defaultValue=""
      type="number"
    />
  </label>
  <br />
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    Сохранить
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    Отмена
  </button>
</div>
`;

exports[`MobileCompany фильтр: показать активных 2`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompany__header section"
  >
    <div
      className="MobileCompany__filter section"
    >
      <button
        className="btn"
        disabled={false}
        onClick={[Function]}
      >
        Все
      </button>
      <button
        className="btn"
        disabled={false}
        onClick={[Function]}
      >
        Активные
      </button>
      <button
        className="btn"
        disabled={false}
        onClick={[Function]}
      >
        Заблокированные
      </button>
    </div>
  </div>
  <table>
    <thead>
      <tr>
        <td>
          Фамилия
        </td>
        <td>
          Имя
        </td>
        <td>
          Отчество
        </td>
        <td>
          Баланс
        </td>
        <td>
          Статус
        </td>
        <td>
          Редактировать
        </td>
        <td>
          Удалить
        </td>
      </tr>
    </thead>
    <tbody>
      <tr
        className="MobileClient"
      >
        <td>
          Маховицкий
        </td>
        <td>
          Эрнест
        </td>
        <td>
          Афанасиевич
        </td>
        <td>
          10.55
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
          >
            Редактировать
          </button>
        </td>
        <td>
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    Добавить нового
  </button>
</div>
`;
